node/src/main/scala/coop/rchain/node/web/Transaction.scala

```diff
@@ -0,0 +1,240 @@
+import coop.rchain.store.{KeyValueStoreManager, KeyValueTypedStore}
+import coop.rchain.shared.syntax._
+import scodec.Codec
+import scodec.codecs.utf8
+
+import scala.collection.concurrent.TrieMap
+import scala.language.higherKinds
+
+final case class Transaction(
+    fromAddr: String,
+    toAddr: String,
+    amount: Long,
+    retUnforgeable: Par,
+    failReason: Option[String]
+)
+
+sealed trait TransactionType
+final case class PreCharge(deployId: String)       extends TransactionType
+final case class UserDeploy(deployId: String)      extends TransactionType
+final case class Refund(deployId: String)          extends TransactionType
+sealed trait SystemTransaction                     extends TransactionType
+final case class CloseBlock(blockHash: String)     extends SystemTransaction
+final case class SlashingDeploy(blockHash: String) extends SystemTransaction
+
+final case class TransactionInfo(transaction: Transaction, transactionType: TransactionType)
+
+final case class TransactionResponse(data: Seq[TransactionInfo])
+
+trait TransactionAPI[F[_]] {
+  def getTransaction(blockHash: BlockHash): F[Seq[TransactionInfo]]
+}
+
+/**
+  * This API is totally based on how RevVault.rho is written. If the `RevVault.rho` is re-written or changed,
+  * this API might end up with useless.
+  */
+final case class TransactionAPIImpl[F[_]: Concurrent](
+    blockReportAPI: BlockReportAPI[F],
+    // The transferUnforgeable can be retrieved based on the deployer and the timestamp of RevVault.rho
+    // in the genesis ceremony.
+    transferUnforgeable: Par
+) extends TransactionAPI[F] {
+  // Helper which accepts TransactionType constructor
+  private def makeDeployTransaction(
+      d: DeployInfoWithEventData,
+      report: SingleReport,
+      txCtor: String => TransactionType
+  ) = findTransactions(report).map(t => TransactionInfo(t, txCtor(d.deployInfo.get.sig)))
+
+  def getTransaction(blockHash: BlockHash): F[Seq[TransactionInfo]] = {
+    val blockHashStr = Base16.encode(blockHash.toByteArray)
+    blockReportAPI.blockReport(blockHash, forceReplay = false).map {
+      case Right(b) =>
+        val userTransactions = b.deploys.flatMap(d => {
+          d.report
+            .zip(Seq(PreCharge, UserDeploy, Refund))
+            .flatMap {
+              case (report, txCtor) =>
+                makeDeployTransaction(d, report, txCtor)
+            }
+        })
+        val systemDeployTransactions = b.systemDeploys.flatMap { s =>
+          // system doesn't get precharge and refund , so it would always get one
+          val transactions = findTransactions(s.report.head)
+          s.systemDeploy.get.systemDeploy.value match {
+            case SlashSystemDeployDataProto(_, _) =>
+              transactions.map(
+                t => TransactionInfo(t, SlashingDeploy(blockHashStr))
+              )
+            case CloseBlockSystemDeployDataProto() =>
+              transactions.map(t => TransactionInfo(t, CloseBlock(blockHashStr)))
```

Comment:
Here the same code is duplicated on both branches, only `SlashingDeploy` or `CloseBlock` are different. So by DRY principle we can first select what is different and have transactions transform specified only once.
```scala
val txCtor = s.systemDeploy.get.systemDeploy.value match {
  case SlashSystemDeployDataProto(_, _)  => SlashingDeploy
  case CloseBlockSystemDeployDataProto() => CloseBlock
}
transactions.map(t => TransactionInfo(t, txCtor(blockHashStr)))
```